### PR TITLE
Create urlsplit filter

### PR DESCRIPTION
**Pull request by samdoran**
##### SUMMARY

Create `urlsplit` filter, documentation, and tests.

Supersedes #20495

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME
filters

##### ANSIBLE VERSION

```
2.4
```

### CODE REVIEW

**Comment by abadger** [comment]
shippable failure passed on second run.  Due to windows timeout so unrelated to this code.

+1 to merge.

**Comment by bcoca** [comment]
looks good, but i would still skip the parameter part, you always do the full work anyways, let the user just choose the key from the dict

File: lib/ansible/plugins/filter/urlsplit.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    split_url = dict(
+        fragment=results.fragment,
+        hostname=results.hostname,
+        password=results.password,
+        path=results.path,
+        port=results.port,
+        query=results.query,
+        scheme=results.scheme,
+        username=results.username
+    )
```
Comment (bcoca) [inline lib/ansible/plugins/filter/urlsplit.py]: you could just use this function instead https://github.com/ansible/ansible/blob/devel/lib/ansible/utils/helpers.py#L37

**Comment by samdoran** [comment]
@bcoca I changed this to use the helper method and eliminate the query. While it works, I don't like the interface, unless I'm missing something.

My playbook: 

```yaml
- hosts: localhost
  connection: local
  gather_facts: no
  become: no

  vars:
    url: http://mary:MySecret@www.acme.com:9000/dir/index.html?query=term#fragment

  tasks:
    - debug:
        var: (url | urlsplit)
```

Which returns a rather hard to read dict with extra fields that aren't really useful:
```
ok: [localhost] => {
    "(url | urlsplit)": "{'count': <built-in method count of SplitResult object at 0x103424440>, 'geturl': <bound method SplitResult.geturl of SplitResult(scheme=u'http', netloc=u'mary:MySecret@www.acme.com:9000', path=u'/dir/index.html', query=u'query=term', fragment=u'fragment')>, 'password': u'MySecret', 'netloc': u'mary:MySecret@www.acme.com:9000', 'query': u'query=term', 'index': <built-in method index of SplitResult object at 0x103424440>, '_make': <bound method type._make of <class 'urlparse.SplitResult'>>, 'scheme': u'http', 'hostname': u'www.acme.com', 'username': u'mary', '_replace': <bound method SplitResult._replace of SplitResult(scheme=u'http', netloc=u'mary:MySecret@www.acme.com:9000', path=u'/dir/index.html', query=u'query=term', fragment=u'fragment')>, '_asdict': <bound method SplitResult._asdict of SplitResult(scheme=u'http', netloc=u'mary:MySecret@www.acme.com:9000', path=u'/dir/index.html', query=u'query=term', fragment=u'fragment')>, 'fragment': u'fragment', 'path': u'/dir/index.html', '_fields': ('scheme', 'netloc', 'path', 'query', 'fragment'), 'port': 9000}"
}
```

I managed to clean it up a bit by adding a few more filters to the helper function:
```
if not key.startswith(('__', '_', 'count', 'geturl', 'index'))
```

But I have a feeling there's a better way to prune the list, and it shouldn't be done in the helper function but in the filter.


--------

The other thing I don't like is to get a single field, it requires parentheses around the expression:
```
var: (url | urlsplit)['port']

...

ok: [localhost] => {
    "(url | urlsplit).port": "9000"
}
```

I'm thinking it's a nicer UI to pass a query to the filter rather than having to wrap the whole thing in parenthesis:
```
var: url | urlsplit('port')
```

Giving the option to do it either way is the best, I think. So if someone wants to pass a query, they get a specific field. If they want to use the dictionary notation, they can.

**Comment by bcoca** [comment]
I didn't expect that many fields in the object, did not see those in docs (i'll fixup utility function so you don't need to duplicate) ... I think the parameter is more work than it's worth, but I'm not going to press on it if others are fine with it.

**Pull request merged**